web/api/v1/api.go

```diff
@@ -927,6 +977,82 @@ func (api *API) remoteRead(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func (api *API) streamedChunkedRemoteRead(ctx context.Context, w http.ResponseWriter, req *prompb.ReadRequest) {
+	externalLabels := api.config().GlobalConfig.ExternalLabels.Map()
+	for i, query := range req.Queries {
+		filteredMatchers, err := filterExtLabelsFromMatchers(query.Matchers, externalLabels)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		querier, err := api.Queryable.Querier(ctx, query.StartTimestampMs, query.EndTimestampMs)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		defer func() {
+			if err := querier.Close(); err != nil {
+				level.Warn(api.logger).Log("msg", "error on querier close", "err", err.Error())
+			}
+		}()
+
+		var selectParams *storage.SelectParams
+		if query.Hints != nil {
+			selectParams = &storage.SelectParams{
+				Start: query.Hints.StartMs,
+				End:   query.Hints.EndMs,
+				Step:  query.Hints.StepMs,
+				Func:  query.Hints.Func,
+			}
+		}
+
+		// TODO(bwplotka): Change interface / find a way to select chunks.
+		set, _, err := querier.Select(selectParams, filteredMatchers...)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		// Add external labels back in, in sorted order.
+		sortedExternalLabels := make([]prompb.Label, 0, len(externalLabels))
+		for name, value := range externalLabels {
+			sortedExternalLabels = append(sortedExternalLabels, prompb.Label{
+				Name:  string(name),
+				Value: string(value),
+			})
+		}
+		sort.Slice(sortedExternalLabels, func(i, j int) bool {
+			return sortedExternalLabels[i].Name < sortedExternalLabels[j].Name
+		})
+
+		w.Header().Set("Content-Type", "application/x-streamed-protobuf; proto=prometheus.ChunkedReadResponse")
+		// TODO(bwplotka): Should we use snappy? benchmark to see.
+		// w.Header().Set("Content-Encoding", "snappy")
+
+		f, ok := w.(http.Flusher)
+		if !ok {
+			http.Error(w, "internal http.ResponseWriter does not implement http.Flusher interface", http.StatusInternalServerError)
+			return
+		}
```

Comment:
When is this the case?

Reply:
When someone refactors other part of the code (wrap it with something that does not implement this).

We have lots of middlewares so it is super likely (: With this we can catch that in unit tests.